File: test/extensions/arguments/with_type_info_code/cpp/TypeInfoStringExpressionsTest.cpp
#include "gtest/gtest.h"
#include "type_info_string_expressions/TypeInfoStringExpressions.h"
#include "zserio/SerializeUtil.h"

using namespace zserio::literals;

namespace type_info_string_expressions
{

using allocator_type = TypeInfoStringExpressions::allocator_type;
using ITypeInfo = zserio::IBasicTypeInfo<allocator_type>;

class TypeInfoStringExpressionsTest : public ::testing::Test
{
protected:
    static void fillData(TypeInfoStringExpressions& typeInfoStringExpressions)
    {
        TestChoice& choiceField = typeInfoStringExpressions.getChoiceField();
        choiceField.setStructField(TestStruct());
        TestStruct& structField = choiceField.getStructField();
        structField.getArrayField().resize(LENGTHOF_LITERAL * 2);
        structField.setDynBitField(0);
        structField.setBitField(0);
        structField.setEnumField(TestEnum::ONE);
        structField.setBitmaskField(TestBitmask::Values::READ);
    }

private:
    static constexpr size_t LENGTHOF_LITERAL = 7;
};

constexpr size_t TypeInfoStringExpressionsTest::LENGTHOF_LITERAL;

TEST_F(TypeInfoStringExpressionsTest, writeRead)
{
    TypeInfoStringExpressions typeInfoStringExpressions;
    fillData(typeInfoStringExpressions);

    auto bitBuffer = zserio::serialize(typeInfoStringExpressions);
    TypeInfoStringExpressions readTypeInfoStringExpressions =
            zserio::deserialize<TypeInfoStringExpressions>(bitBuffer);

    ASSERT_EQ(typeInfoStringExpressions, readTypeInfoStringExpressions);
}

TEST_F(TypeInfoStringExpressionsTest, typeInfo)
{
    const ITypeInfo& typeInfo = TypeInfoStringExpressions::typeInfo();

    // choiceField
    const auto& choiceFieldInfo = typeInfo.getFields()[0];
    ASSERT_EQ("::zserio::makeStringView(\"literal\").size()"_sv, choiceFieldInfo.typeArguments[0]);

    // TestChoice
    const auto& choiceTypeInfo = choiceFieldInfo.typeInfo;
    ASSERT_EQ("getSelector() + ::zserio::makeStringView(\"literal\").size()"_sv, choiceTypeInfo.getSelector());
    ASSERT_EQ("0 + ::zserio::makeStringView(\"literal\").size()"_sv,
            choiceTypeInfo.getCases()[0].caseExpressions[0]);

    // structField
    const auto& structFieldInfo = choiceTypeInfo.getFields()[1];
    ASSERT_EQ("::zserio::makeStringView(\"literal\").size()"_sv, structFieldInfo.typeArguments[0]);

    // TestStruct
    const auto& structTypeInfo = structFieldInfo.typeInfo;

    // arrayField
    const auto& arrayFieldInfo = structTypeInfo.getFields()[0];
    ASSERT_EQ("getParam() + ::zserio::makeStringView(\"literal\").size()"_sv, arrayFieldInfo.arrayLength);
    ASSERT_EQ("getArrayField().size() > ::zserio::makeStringView(\"literal\").size()"_sv,
            arrayFieldInfo.constraint);

    // dynBitField
    const auto& dynBitFieldInfo = structTypeInfo.getFields()[1];
    ASSERT_EQ("::zserio::makeStringView(\"literal\").size()"_sv, dynBitFieldInfo.typeArguments[0]);
    ASSERT_EQ("getParam() + ::zserio::makeStringView(\"literal\").size() != 0"_sv,
            dynBitFieldInfo.optionalCondition);

    // bitField
    const auto& bitFieldInfo = structTypeInfo.getFields()[2];
    ASSERT_EQ("::zserio::makeStringView(\"literal\").size() + 1"_sv, bitFieldInfo.alignment);

    // enumField
    const auto& enumFieldInfo = structTypeInfo.getFields()[3];

    // TestEnum
    const auto& enumTypeInfo = enumFieldInfo.typeInfo;
    ASSERT_EQ("::zserio::makeStringView(\"literal\").size()"_sv, enumTypeInfo.getUnderlyingTypeArguments()[0]);

    // bitmaskField
    const auto& bitmaskFieldInfo = structTypeInfo.getFields()[4];

    // TestBitmask
    const auto& bitmaskTypeInfo = bitmaskFieldInfo.typeInfo;
    ASSERT_EQ(
            "::zserio::makeStringView(\"literal\").size()"_sv, bitmaskTypeInfo.getUnderlyingTypeArguments()[0]);
}

} // namespace type_info_string_expressions

File: test/extensions/arguments/with_type_info_code/python/TypeInfoStringExpressionsTest.py
import unittest
import zserio

from zserio.typeinfo import MemberAttribute, TypeAttribute

from testutils import getZserioApi


class TypeInfoStringExpressionsTest(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.api = getZserioApi(__file__, "type_info_string_expressions.zs", extraArgs=["-withTypeInfoCode"])

    def testTypeInfo(self):
        typeInfoStringExpressions = self._createData()
        typeInfo = self.api.TypeInfoStringExpressions.type_info()

        # choiceField
        choiceField = typeInfoStringExpressions.choice_field
        choiceFieldInfo = typeInfo.attributes[TypeAttribute.FIELDS][0]
        self.assertEqual(
            self.LENGTHOF_LITERAL,
            choiceFieldInfo.attributes[MemberAttribute.TYPE_ARGUMENTS][0](typeInfoStringExpressions, 0),
        )

        # TestChoice
        choiceTypeInfo = choiceFieldInfo.type_info
        choiceSelector = choiceTypeInfo.attributes[TypeAttribute.SELECTOR]
        self.assertEqual(2 * self.LENGTHOF_LITERAL, choiceSelector(choiceField))

        caseInfo0 = choiceTypeInfo.attributes[TypeAttribute.CASES][0]
        self.assertEqual(self.LENGTHOF_LITERAL, caseInfo0.case_expressions[0]())

        # structField
        structField = choiceField.struct_field
        structFieldInfo = choiceTypeInfo.attributes[TypeAttribute.FIELDS][1]
        self.assertEqual(
            self.LENGTHOF_LITERAL, structFieldInfo.attributes[MemberAttribute.TYPE_ARGUMENTS][0](choiceField, 0)
        )

        # TestStruct
        structTypeInfo = structFieldInfo.type_info

        # arrayField
        arrayFieldInfo = structTypeInfo.attributes[TypeAttribute.FIELDS][0]
        self.assertEqual(
            2 * self.LENGTHOF_LITERAL, arrayFieldInfo.attributes[MemberAttribute.ARRAY_LENGTH](structField)
        )
        self.assertTrue(arrayFieldInfo.attributes[MemberAttribute.CONSTRAINT](structField))

        # dynBitField
        dynBitFieldInfo = structTypeInfo.attributes[TypeAttribute.FIELDS][1]
        self.assertEqual(
            self.LENGTHOF_LITERAL, dynBitFieldInfo.attributes[MemberAttribute.TYPE_ARGUMENTS][0](structField, 0)
        )
        self.assertTrue(dynBitFieldInfo.attributes[MemberAttribute.OPTIONAL](structField))

        # bitField
        bitFieldInfo = structTypeInfo.attributes[TypeAttribute.FIELDS][2]
        self.assertEqual(self.LENGTHOF_LITERAL + 1, bitFieldInfo.attributes[MemberAttribute.ALIGN]())

        # enumField
        enumFieldInfo = structTypeInfo.attributes[TypeAttribute.FIELDS][3]

        # TestEnum
        enumTypeInfo = enumFieldInfo.type_info
        self.assertEqual(
            self.LENGTHOF_LITERAL, enumTypeInfo.attributes[TypeAttribute.UNDERLYING_TYPE_ARGUMENTS][0]()
        )

        # bitmaskField
        bitmaskFieldInfo = structTypeInfo.attributes[TypeAttribute.FIELDS][4]

        # TestBitmask
        bitmaskTypeInfo = bitmaskFieldInfo.type_info
        self.assertEqual(
            self.LENGTHOF_LITERAL, bitmaskTypeInfo.attributes[TypeAttribute.UNDERLYING_TYPE_ARGUMENTS][0]()
        )

    def testWriteRead(self):
        typeInfoStringExpressions = self._createData()
        bitBuffer = zserio.serialize(typeInfoStringExpressions)
        readTypeInfoStringExpressions = zserio.deserialize(self.api.TypeInfoStringExpressions, bitBuffer)
        self.assertEqual(typeInfoStringExpressions, readTypeInfoStringExpressions)

    def _createData(self):
        return self.api.TypeInfoStringExpressions(
            self.api.TestChoice(
                self.LENGTHOF_LITERAL,
                struct_field_=self.api.TestStruct(
                    self.LENGTHOF_LITERAL,  # param
                    [0] * self.LENGTHOF_LITERAL * 2,  # array
                    0,  # dynBitField
                    0,  # bitField
                    self.api.TestEnum.ONE,
                    self.api.TestBitmask.Values.READ,
                ),
            )
        )

    LENGTHOF_LITERAL = 7
